src/types.ts

```diff
@@ -3382,6 +3383,7 @@ export class ZodRecord<
         value: valueType._parse(
           new ParseInputLazyPath(ctx, ctx.data[key], ctx.path, key)
         ),
+        alwaysSet: key in ctx.data,
```

Comment:
We just got `key` from `for (const key in ctx.data)`, so this will always be `true`, right?

Reply:
Yes, should be `alwaysSet: true` here.
Also currently only `ParseStatus.mergeObjectSync` supports `alwaysSet`, so I updated the async one as well `ParseStatus.mergeObjectAsync`(and added a test for it).